### ISSUE TITLE
http dependency telemetry is not correlated correctly with RequestTelemetry.

### ISSUE THREAD

**Issue by hallatore** [Microsoft/ApplicationInsights-dotnet-server#112]
Ref: https://github.com/Microsoft/ApplicationInsights-dotnet/issues/268

None of my http requests are correlated with it request. It seems that maybe the async nature of it breaks the correlation.

**Comment by hallatore** [comment]
Here is async HttpClient vs. legacy HttpWebRequest. Notice the missing tags in the first item.

```JavaScript
{ // HttpClient
    "name": "Microsoft.ApplicationInsights.Dev.7bda6b4408cf4392bfff847d536e842b.RemoteDependency",
    "time": "2016-08-01T16:35:30.2502432Z",
    "iKey": "7bda6b44-08cf-4392-bfff-847d536e842b",
    "tags": {
        "ai.device.roleInstance": "WIN-UV1B28GR3IV",
        "ai.internal.sdkVersion": "rddf: 2.1.0.363"
    },
    "data": {
        "baseType": "RemoteDependencyData",
        "baseData": {
            "ver": 2,
            "name": "http://www.vg.no/",
            "id": "N5lM+k2gf/A=",
            "value": 17.5045,
            "resultCode": "200",
            "dependencyKind": 1,
            "success": true,
            "properties": {
                "DeveloperMode": "true"
            }
        }
    }
}, 
{ // HttpWebRequest
    "name": "Microsoft.ApplicationInsights.Dev.7bda6b4408cf4392bfff847d536e842b.RemoteDependency",
    "time": "2016-08-01T16:35:30.2992553Z",
    "iKey": "7bda6b44-08cf-4392-bfff-847d536e842b",
    "tags": {
        "ai.operation.parentId": "cE7yWcVJW9Y=",
        "ai.device.roleInstance": "WIN-UV1B28GR3IV",
        "ai.operation.name": "GET Home/Index",
        "ai.operation.id": "cE7yWcVJW9Y=",
        "ai.internal.sdkVersion": "rddf: 2.1.0.363"
    },
    "data": {
        "baseType": "RemoteDependencyData",
        "baseData": {
            "ver": 2,
            "name": "http://www.dagbladet.no/",
            "id": "UiljQXRjHuY=",
            "value": 28.6517,
            "resultCode": "200",
            "dependencyKind": 1,
            "success": true,
            "properties": {
                "DeveloperMode": "true"
            }
        }
    }
}
```

**Comment by hallatore** [comment]
The same thing happens if you wrap HttpWebRequest in a `await Task.Run(...);`

I think it maybe has something to do with CallContext not "surviving" an async/await?
Ref: https://github.com/Microsoft/ApplicationInsights-dotnet/blob/develop/src/Core/Managed/Operation.CC.Shared/Extensibility/Implementation/CallContextHelpers.cs#L27

**Comment by abaranch** [comment]
Currently it is by design. We are looking into ways how we can improve it.
Duplicate: Microsoft/ApplicationInsights-dotnet-server#106

**Comment by hallatore** [comment]
How is this "by design" and not a bug? Is it "by design" as in: not possible to get it to work with the current pipeline?
What ways are you looking into improving this in the future?

**Comment by abaranch** [comment]
Using HttpContext assumes that correlation does not work in async scenarios. So it never worked and it is a known limitation right now. 
A workaround would be to use Operations API.
We are looking into a way to pass required information in cases when we have a profiler (dependencies collected when you have StatusMonitor installed) sticking it into parameters to which we have access like IAsyncResult.
It is on the list of things to investigate but there is no ETA.

**Comment by david-jarman** [comment]
@abaranch, how do you propose to use Operations API as a workaround for async HTTP calls? 

I'm running into the same issue for several types of dependencies:

1. Dependencies made in a different thread (Via HostingEnvironment.QueueBackgroundWorkItem) 
2. Async HTTP calls like the one described in this issue.
3. Async calls to blob storage that we use Task.WhenAll(List<Task> blobCalls) to parallelize downloading the contents of the blobs.

I solved #1 by using the Operations API improvements in 2.3.0-beta. 
How would it be possible to solve 2 and 3 using this api? You don't have access to the thread like you do in QueueBackgroundWorkItem.

Thanks!
